server/src/main/java/org/opensearch/index/query/BaseQueryRewriteContext.java

```diff
@@ -90,21 +91,27 @@ public QueryShardContext convertToShardContext() {
      * from an index.
      */
     public void registerAsyncAction(BiConsumer<Client, ActionListener<?>> asyncAction) {
-        asyncActions.add(asyncAction);
+        asyncActionsRef.updateAndGet(list -> {
+            List<BiConsumer<Client, ActionListener<?>>> newList = new ArrayList<>(list);
+            newList.add(asyncAction);
+            return newList;
+        });
```

Comment:
I don't think we need to copy the list here; this gives us no value over a `CopyOnWriteArrayList` which we wouldn't need the atomic reference for.   Can't we simply just update the list in-place but wrapped in the atomic reference for thread safety?  (CC @msfroh to confirm this suggestion):
```suggestion
        asyncActionsRef.updateAndGet(list -> {
            list.add(asyncAction);
            return list;
        });
```

Reply:
Related, I think if you declare the initial list as a `Collections.synchronizedList(ArrayList<>)` then this whole line could be even simpler:
```java
asyncActionsRef.get().add(asyncAction);
```
Not sure about performance tradeoffs here, and I would like a second set of eyes on this suggestion as well.

Reply:
I tried, if not returning new list, I am having test failure, so it still cause racing condition during registration. 

```
BaseQueryRewriteContextTests > testRacingConditionFixed FAILED
    java.lang.AssertionError: expected:<5000> but was:<4874>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.opensearch.index.query.BaseQueryRewriteContextTests.testRacingConditionFixed(BaseQueryRewriteContextTests.java:167)

BaseQueryRewriteContextTests > testConcurrentRegistrationAndExecution FAILED
    java.lang.AssertionError: expected:<10000> but was:<7802>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.opensearch.index.query.BaseQueryRewriteContextTests.testConcurrentRegistrationAndExecution(BaseQueryRewriteContextTests.java:108)

```

Reply:
Hmm... if we need to copy the whole list anyway, I think I like @dbwiddis's suggestion of using `CopyOnWriteArrayList` over my `AtomicReference` suggestion.